### ISSUE TITLE
feature request: cond() for sparse matrices

### ISSUE THREAD

**Issue by alsam** [JuliaLang/julia#6485]
Julia do have condition number computation `cond` algorithm for dense matrices but lacks it for sparse matrices:

```jlcon
julia> x = sprandn(7,7,0.3);

julia> cond(x)
ERROR: no method cond(SparseMatrixCSC{Float64,Int64})

julia> y = rand(7,7)
7x7 Array{Float64,2}:
 0.105457  0.820211    0.526384    0.257885  0.0951214   0.619188  0.11465  
 0.328247  0.00401575  0.417114    0.747672  0.167338    0.132277  0.298542 
 0.624044  0.564439    0.00227799  0.883262  0.335541    0.779947  0.93982  
 0.231413  0.63107     0.523263    0.823538  0.452903    0.129218  0.0104342
 0.562113  0.536426    0.167322    0.565356  0.00539442  0.477658  0.522652 
 0.868369  0.279572    0.101924    0.435134  0.483316    0.47326   0.655955 
 0.492717  0.204044    0.230936    0.113277  0.912932    0.393085  0.269244 

julia> cond(y)
592.9028870805738
```

Thanks!

**Comment by gwhowell** [comment]
For condition estimation, another possible algorithm is being championed by Sivan Toledo 
Here's a link. 
http://www.eecs.berkeley.edu/~odedsc/seminars/scmc-seminar-spring2013_files/Sivan-Toledo-Slides.pdf

Allows estimate of conditioning for non square matrices ?

**Comment by gwhowell** [comment]
http://www.eecs.berkeley.edu/~odedsc/seminars/scmc-seminar-spring2013_files/Sivan-Toledo-Slides.pdf

interesting -- much faster than computing the SVD, slow sometimes .. he does survey other methods ..

**Comment by jiahao** [comment]
@gwhowell Thanks, these look like pretty interesting slides.

The basic idea has been around for awhile - Parlett's and Saad's books (amongst others) have described how you can get estimates of eigenvalues (and hence the condition number) from studying the Lanczos vectors that underly iterative solvers. The new idea here seems to be his use of the forward error to study the convergence rate of LSQR. Normally the forward error is unknown, but the idea here is to simply pick a problem with a known solution and compute the Rayleigh quotient for the smallest singular value when it is converged.

The usual caveats of iterative methods apply. The main problem (which is acknowledged in the slides but not addressed) is that the convergence rate depends strongly on how well-separated the singular value you are interested in is from the other singular values. Many matrices do not have good σ₂ - σ₁ spectral gaps, and so estimating the condition number remains a difficult problem in the general case.

This could be a feature request for IterativeSolvers.jl

**Comment by poulson** [comment]
For what it's worth, the de facto method right now for one-norm estimation is the block variant of the mentioned Hager-Higham algorithm: http://epubs.siam.org/doi/abs/10.1137/S0895479899356080?journalCode=sjmael

**Comment by gwhowell** [comment]
Hi Jack,
I think the  Hager-Higham algorithm you mention is the one used in octave
and matlab -- the operator is called
condest

Here's the octave documentation
'condest' is a function from the file
/usr/share/octave/3.6.4/m/linear-algebra/condest.m

 -- Function File: condest (A)
 -- Function File: condest (A, T)
 -- Function File: [EST, V] = condest (...)
 -- Function File: [EST, V] = condest (A, SOLVE, SOLVE_T, T)
 -- Function File: [EST, V] = condest (APPLY, APPLY_T, SOLVE, SOLVE_T,
          N, T)

     Estimate the 1-norm condition number of a matrix A using T test
     vectors using a randomized 1-norm estimator.  If T exceeds 5, then
     only 5 test vectors are used.

     If the matrix is not explicit, e.g., when estimating the condition
     number of A given an LU factorization, 'condest' uses the following
     functions:

     APPLY
          'A*x' for a matrix 'x' of size N by T.

    APPLY_T
          'A'*x' for a matrix 'x' of size N by T.

    SOLVE
          'A \ b' for a matrix 'b' of size N by T.

     SOLVE_T
          'A' \ b' for a matrix 'b' of size N by T.

     The implicit version requires an explicit dimension N.

     'condest' uses a randomized algorithm to approximate the 1-norms.

     'condest' returns the 1-norm condition estimate EST and a vector V
     satisfying 'norm (A*v, 1) == norm (A, 1) * norm (V, 1) / EST'.
     When EST is large, V is an approximate null vector.

     References:
        * N.J. Higham and F. Tisseur, 'A Block Algorithm for Matrix
          1-Norm Estimation, with an Application to 1-Norm
          Pseudospectra'.  SIMAX vol 21, no 4, pp 1185-1201.
          <http://dx.doi.org/10.1137/S0895479899356080>

        * N.J. Higham and F. Tisseur, 'A Block Algorithm for Matrix
          1-Norm Estimation, with an Application to 1-Norm
          Pseudospectra'.  <http://citeseer.ist.psu.edu/223007.html>

     See also: cond, norm, onenormest

Regards,
Gary Howell

**Comment by StefanKarpinski** [comment]
Please don't quote the copyrighted documentation of other projects here. It might qualify as fair use, but it's better not to risk any copyright infringement.

**Comment by poulson** [comment]
@gwhowell Agreed, that's what I was implying when I said "de facto". From what I recall from my last brief conversation with Higham, MATLAB defaults to setting the blocksize to 2.

**Comment by jiahao** [comment]
> Please don't quote the copyrighted documentation of other projects here. It might qualify as fair use, but it's better not to risk any copyright infringement.

attn: @gwhowell - we've asked you on previous occasions not to do this.

**Comment by poulson** [comment]
@StefanKarpinski @jiahao I assume that quoting documentation from permissively licensed (e.g., New BSD) is okay?

**Comment by tkelman** [comment]
Better to just link to things in general, if they're more than a few lines

**Comment by mfasi** [comment]
I implemented the Hager-Higham algorithm in Julia (a tentative [gist](https://gist.github.com/mfasi/bfa2664fbc385470dda3)) and would be happy to add it to base, but I'm not sure what the best way to do this will be. If I run `cond(A)` with a `A = sprandn(n,n,d)` I get the following

```javascript
julia> cond(A,2)
ERROR: MethodError: `svdvals!` has no method matching svdvals!(::Base.SparseMatrix.SparseMatrixCSC{Float64,Int64})
 in cond at linalg/dense.jl:494
 in cond at linalg/dense.jl:493

julia> cond(A,1)
ERROR: MethodError: `cond` has no method matching cond(::Base.SparseMatrix.UMFPACK.UmfpackLU{Float64,Int64}, ::Int64)
Closest candidates are:
  cond(::Number, ::Any)
  cond{T<:Union{Complex{Float32},Float64,Float32,Complex{Float64}},S}(::Base.LinAlg.LowerTriangular{T<:Union{Complex{Float32},Float64,Float32,Complex{Float64}},S}, ::Real)
  cond{T<:Union{Complex{Float32},Float64,Float32,Complex{Float64}},S}(::Base.LinAlg.UnitLowerTriangular{T<:Union{Complex{Float32},Float64,Float32,Complex{Float64}},S}, ::Real)
  ...
 in cond at linalg/dense.jl:499

julia> cond(A,Inf)
ERROR: MethodError: `cond` has no method matching cond(::Base.SparseMatrix.UMFPACK.UmfpackLU{Float64,Int64}, ::Float64)
Closest candidates are:
  cond(::Number, ::Any)
  cond{T<:Union{Complex{Float32},Float64,Float32,Complex{Float64}},S}(::Base.LinAlg.LowerTriangular{T<:Union{Complex{Float32},Float64,Float32,Complex{Float64}},S}, ::Real)
  cond{T<:Union{Complex{Float32},Float64,Float32,Complex{Float64}},S}(::Base.LinAlg.UnitLowerTriangular{T<:Union{Complex{Float32},Float64,Float32,Complex{Float64}},S}, ::Real)
  ...
 in cond at linalg/dense.jl:499
```

which I think is more misleading than the original `ERROR: no method cond(SparseMatrixCSC{Float64,Int64})`.

I think what we should have is:

* A `condest` method to estimate the 1-norm condition number for both dense and sparse matrices
* A `cond` method for `AbstractMatrix` (in generic.jl rather than dense.jl, maybe) which computes `cond(A)` if `A` is dense and returns a message to help the user choose between `cond(full(A))` and `condest(A)`.

Does it sound reasonable? In case it does I'll be happy to adapt the gist to base and open a pull request. Comments on the code would be appreciated as well.

**Comment by tkelman** [comment]
I've changed my mind since April 2014, while I still think this code would be really good to have cleaned up and easily available somewhere, I'm not sure it absolutely needs to be in base unless it's important enough to be a dependency of all other code written in the language. Especially now with package precompilation, what about making a package for it, or contributing it to somewhere like https://github.com/andreasnoack/LinearAlgebra.jl (which is not yet registered, but probably will be sooner or later)?

**Comment by mfasi** [comment]
@tkelman Ok, thank you, I'll clean it up and then see where to put it. I might need it at some point to implement one of the functions in #5840, but I'm not sure yet.
However, I think that `cond(A)` should return an informative message rather than an apparently unrelated error when `A` is a `SparseMatrix`. Maybe it would be enough to do something similar to what `inv(A)` does?

**Comment by andreasnoack** [comment]
As long as we haven't moved larger pieces of the sparse functionality out of base I think the condition number estimates for sparse matrices belong in base. It should just be called `cond` and without the `est` part. The LAPACK condition number subroutine returns estimates and you could also argue that the computed spectral condition number is an estimate (just with a small error).

**Comment by tkelman** [comment]
Let's be honest, condition estimates are a pretty niche feature. And I don't think we should call it `cond` if it's an estimate.

**Comment by andreasnoack** [comment]
Improvements to the documentation are always welcome and I agree that it would great to clarify how the different condition numbers are computed. Preferably with links to the relevant sources/papers.

**Comment by gwhowell** [comment]
Along with a link to the relevant papers, the documentation should point
out what condition number is being estimated.  For the Higham Hager
algorithm, that's L_1 condition number for example (don't know that the
current dense cond() is estimating that same number, if not, that would be
argument for using two different names for the sparse and dense cases)

**Comment by andreasnoack** [comment]
In general, I don't think we should have different functions for different algorithms for the same quantity. We should pick a good default and document it. If we'd like to give choices (which I'm considering for for eigen and singular values decompositions) I think we should use a keyword argument to select the algorithm.

**Comment by mfasi** [comment]
In this case, `cond(A, p)` where p = 1, 2 or `Inf` computes the p-condition number of a matrix, the default being p = 2. The Hager-Higham algorithm estimates the 1 (and thus `Inf`) condition number, the case p = 2 can be implemented as the power method on `A' * A`.

**Comment by andreasnoack** [comment]
I wouldn't bother with the spectral condition number for now. It would be great just to have 1 and Inf. The smallest singular value would be pretty inaccurate if calculated with `A'A`. In principle, we should be able to calculate the spectral condition number with `svds`, but it appears that right now our implementation only allows us to compute the largest values.

**Comment by andreasnoack** [comment]
Closed by #12467
